Update results for Duktape 2.2

svaarala opened the pull request:
- Add Duktape 2.2 engine.
- Update results for Duktape 2.2.0 release.
- Change `duktape2_0: false` to `duktape_1.0: false` now that Duktape 1.x engine versions have been added; previously results started from Duktape 2.0.0.
- Remove a few redundancies, e.g. the 2.0 line for `duktape1_3: true` and  `duktape_2.0: true`.

svaarala committed:
Add Duktape 2.2 environment

afmenez commented on environments.json:
```diff
@@ -2008,7 +2008,7 @@
     "short": "DUK 2.2",
     "platformtype": "engine",
-    "unstable": true
+    "obsolete": false
```
You could add the release date here.  Actually, I guessed the "retire date" (end of support) for all the previous versions, you could check them and add the retire date for 2.1 too. ;)

svaarala replied:
Hmm, 2.1 will still get at least one round of maintenance fixes so I guess it's not retired yet?

afmenez replied:
Sure.  But the release data from 2.2 is still missing. :)

svaarala replied:
No disagreement there, I'll add it.

afmenez replied:
Actually, your comment on 2.1 still having one round yet makes me doubt of the retire dates I guessed for all other versions.  Can you update or remove them?

afmenez commented on data-esintl.js:
```diff
@@ -22,7 +22,7 @@ exports.tests = [
         opera10_50: false,
         node0_12: true,
-        duktape2_0: false,
+        duktape1_0: false,
```
I did some "result optimization" myself in the beginning, but it is not the way it is supposed to work.  You should keep here the results for the versions that were actually tested, and let the compiler guess the others.  So you are not supposed to remove correct results, only add new ones.

svaarala replied:
I've received a little contradicting feedback on this: previously when I added results for 2.1 that were the same as for 2.0 -- both tested explicitly -- I got feedback that the 2.1 results shouldn't be added if they are not different from 2.0.

afmenez replied:
Interesting.  Maybe it's a major/minor thing then?  Let's hear from the committers then.

svaarala replied:
Also the results for at least Duktape 2.2 (and possibly Duktape 1.3 and 1.4) were added (by someone else) without actually running tests. So some of the 2.2 results were incorrect before the pull. I'm not sure if the 1.x data is correct either (i.e. whether it has actually been tested or is just based on release notes or something like that).

svaarala replied:
@afmenez It seems that the previous pull with 2.2 results was actually from you - can you confirm whether the 1.x results are based on actual test runs or release notes only? If the latter, I could try to run the tests with 1.x to make the results exact?

afmenez replied:
Yes, all the data I added was based on the release notes.  You can sure run the tests if you can and correct everything that is wrong!

svaarala replied:
Ok, I'll do that in a separate pull later on if these results are interesting for old versions :)

afmenez replied:
Actually it would be great if you could add the actual results for 1.x as you did before.  There's a separate PR #1131 that changes the default from "false" to "unknown", and it would affect most of the results for 1.x if they are not explicitly set.

svaarala replied:
Hmm, so the most important thing would be to get 1.0 results for the full test set, because those are inherited forwards automatically.

afmenez replied:
Exactly.

svaarala committed:
Handle master snapshots better in duktape.js

Treat e.g. 2.2.99 as 2.3.0 for testing purposes; it's built from master and
just before the release 2.2.99 should be treated as 2.3 (not 2.2) as far as
results are concerned.

svaarala commented:
Regarding the "retire" dates, I'm not sure what the best (desired) semantics are, as it probably depends on the release policy of each project. But for Duktape releases one clear approach is:
- If the release version has maintenance releases beyond the next minor release, the retire date is the last such maintenance release.
- If the last maintenance release was before the next minor release, the retire date is that next release. In effect the new release replaced the previous one. In between the last maintenance release and the next minor release the last maintenance release was still current (and would have been fixed for any serious bugs).

svaarala committed:
Update data for Duktape 2.2

svaarala commented:
Okay, made the following changes:
- Updated release/retire dates as described above, fixing some 1.x data and adding a release date for 2.2.
- Dropped the previous duktape1_0 data changes, and added only duktape2_2 results for any tests whose result has changed since duktape2_0 or duktape2_1.
- However, the result from duktape2_2 is not added if it hasn't changed since duktape2_1 or duktape2_0; this is consistent with how results for Duktape 2.1 were added (= only changes were recorded).

Squashed and pushed.

afmenez reviewed (approved):
Nice!

ljharb commented on duktape.js:
```diff
@@ -26,6 +26,9 @@ var dukKey = (function () {
     var dukVersion = Number(stdout);
     console.log('Duktape version is: ' + dukVersion);
+    if ((dukVersion % 100) == 99) {
+        dukVersion++;  // Treat e.g. 2.2.99 (built from master) as 2.3.0 for testing
```
Can you elaborate on why this is in here, or why `2.2.99` would be anything more special than the patch version after `2.2.98`?

svaarala replied:
Right after 2.2.0 has been released, Duktape master switches to 2.2.99 which is used until the 2.3.0 release (http://duktape.org/guide.html#versioning.4). While no version is correct for a master build, the .99 patch is used because the build is closer to 2.3.0 than any 2.2.x maintenance version for comparison purposes.


When one runs compat-table/duktape.js with the 2.2.99 release, prior to this change it will appear as Duktape 2.2 and compare against `duktape2_2` results. This is a bit useless because there's no good way to fix the data entries; one would need to alter 2.2 results.

With this change it will compare against `duktape2_3` results and allow one (in practice me) to maintain the to-be-released 2.3.0 results incrementally.

This is not an important change, but before this, I had to manually force dukVersion in duktape.js for testing and maintaining the results incrementally.

ljharb replied:
In that case, duktape is not following semver; it should be using prerelease builds for that. `.100` is no more special than `.1` or `.1000`, and misusing `99` like that means semver-minor changes appear in the 2.2 line after .0.

We'll have to alter the entire way we handle duktape versions inside compat-table unless it changes this policy.

svaarala replied:
Note that builds from master are not releases at all, so Duktape *releases* do follow semver but master builds are not releases.

svaarala replied:
A bit more background, DUK_VERSION is a number of the form 10203 (for 1.2.3) and between releases I guess there are roughly the following options (assuming master build after 2.2.0 but before 2.3.0):
- 2.2.99: current policy, patch level never reaches .99. It's possible to detect an unofficial build from the version number, and 'version >= 2.3.0' check won't catch the unofficial build which is good.
- 2.3.0: this would also be workable, but official release would have to be 2.3.1 to avoid confusion. But this would also break semver: 2.3.0 -> 2.3.1 might entail arbitrary changes (not just patch level changes), e.g. if the 2.3.0 was built right after 2.2.0.
- Alternate releases, e.g. 2.2.0 official release, 2.3.0 reserved for development, 2.4.0 official release, etc. This would be closest to fulfilling semver, but even with this 2.3.x patch level guarantees would be broken because development between minor versions often involves multiple individual changes deserving a minor version bump.

For a numeric version like 1.2.3, with no space for `-alpha` or similar qualifiers, it's not really possible to fulfil semver as far as I can see it, unless one bumps the version literally for every commit going into (unreleased) master, which is very inconvenient. So, that's why the current compromise is to use strict semver for releases only, not manual builds from Duktape master.

ljharb replied:
No, it wouldn't - you'd use "prereleases", which are part of the semver spec and defined for exactly this purpose. You'd do `2.3.0-alpha.0` or `2.3.0-rc.1` etc, and then `2.3.0` when ready.

Using `.99` makes no sense whatsoever.

svaarala replied:
Given that DUK_VERSION is a number of the form major * 10000 + minor * 100 + patch, there's no way to add a suffix like that. The reason it is a number is to allow C programmers to do things like:

```c
#if (DUK_VERSION >= 20300)
/* Use Duktape 2.3 features */
#else
/* Older */
#endif
```

C preprocessor cannot do version comparisons from strings, which is why version number defines are common for C libraries.

ljharb replied:
Since that's the case, that means duktape does not follow semver, and the versioning logic and automated scripts for it in this repo were allowed based primarily on following semver.

svaarala replied:
Does semver require explicit versioning for every *unreleased* commit in a repository?

Because that's what we're talking about here: there are no releases on duktape.org with a .99 patch level, and there won't be any releases like that.

ljharb replied:
Fair point; in that case, let's find a better hack than a ".99" version; or at least let's not let it pollute this repo.

svaarala replied:
Like I explained, that "hack" is the best compromise in practice. Something must go into DUK_VERSION, and 20299 is better than 20300, at least based on user feedback.

But I can remove that check if it makes things better. It just means running duktape.js for a development build without manual tweaks gives nonsensical results. I fail to see why that's better.

ljharb replied:
Personally I find that reserving the `.0` version for this, and only ever releasing `.1` and higher, would be a much cleaner approach than breaking semver just to accomodate deficiencies in C.

svaarala replied:
But that would also break semver: the changes from 2.3.0 -> 2.3.1 would in many cases be more significant than patch level changes?

To clarify, right after 2.2.0 release master would switch to 2.3.0. If you built Duktape right at that time, it would behave like 2.2.0 but have a version number of 2.3.0. In this case the change from 2.3.0 -> 2.3.1 would be the same as 2.2.0 -> 2.3.1.

ljharb replied:
hm, also a fair point. not really sure what to do here.

svaarala replied:
One option is to detect the .99 patch level and refuse to run, indicating the cause: you're running a development version which exists outside the semver version number space and it's not obvious which results you'd want to compare against.

duktape.js could then provide an option to force the version to be used for results purposes. I could then run 2.2.99 as `nodejs duktape.js --force-duk-version 20300` to update the 2.3 results incrementally. This would avoid the need to manually hack the source file.

But I'm fine with whatever really -- I can keep on editing duktape.js manually every time I run the tests, and I doubt many others will run a Duktape development build against the test set.

chicoxyzzy reviewed (approved):
LGTM

ljharb reviewed (dismissed):
I'm still very displeased with duktape not following semver - that it's doing it due to flaws in C++ versioning doesn't excuse it.

svaarala commented:
> let's file a new issue to address how we can deal with duktape not following semver, moving forward.

I quite disagree with this statement as discussed above.

chicoxyzzy commented:
TypeScript and all browsers don't use semver for versioning. Why should it be a blocker?

ljharb commented:
@chicoxyzzy because duktape has special handling inside compat-table based on the assumption that it uses semver (it's also the only engine that has tools to add automated results)

svaarala commented:
@ljharb What special handling are you referring to?

All the Duktape versions listed in environments (i.e., are releases) do follow semver. The .99 versions are only used for builds from master, i.e. non-releases.

ljharb commented:
@svaarala the existence of https://github.com/kangax/compat-table/blob/7db11bd87d2a51e7562a869d3533af01b00d5f4c/duktape.js combined with https://github.com/kangax/compat-table/blob/gh-pages/duktape.js#L43 / https://github.com/kangax/compat-table/blob/gh-pages/duktape.js#L29 - no other engine has special treatment in this way (although many engines have - perhaps problematic - automatic result inheritance based on "not semver")

svaarala commented:
That check makes no semver assumptions as such - it just interprets environments.json as enumerating the Duktape versions in increasing order of "newness".

Results are inherited based on that order as far as I can tell, and as far as I understood the same inheritance approach is used for all engines? IOW, if no result is provided, it is assumed to be the same for a new version described later in environments.json regardless of any versioning conventions?

But, I might have understood the process incorrectly.

svaarala commented:
Also just to be clear, duktape.js does not update the results automatically. When it is executed, it just points out what results don't match the values in the data-* files. The changes to the data files are then made manually just like for other engines (again, as far as I understood).

In this process duktape.js "inherits" results based only on the order of Duktape versions listed in environments.json. There is no inference based on the version number itself, or any intepretation of the version number components. This is done here: https://github.com/kangax/compat-table/blob/gh-pages/duktape.js#L33-L49.

ljharb commented:
You make fair points. Either way, this PR merely needs rebasing to be mergeable.

svaarala committed:
Update HTML files for Duktape 2.2

svaarala commented:
Rebased and pushed.
